### PR TITLE
Indexation improvement

### PR DESCRIPTION
**Pull request by vgoncharenko**
Create Plugins to invalidate StockItem indexes when Stock-to-Source assignments have been changed

### Description
Create Plugins to invalidate StockItem indexes when Stock-to-Source assignments have been changed

### Fixed Issues (if relevant)
1. magento-engcom/msi#68: Create Plugins to invalidate StockItem indexes when Stock-to-Source assignments have been changed

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/Inventory/Indexer/Source.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
+<?php
```
Comment (larsroettig) [inline app/code/Magento/Inventory/Indexer/Source.php]: Please add static PHP static typing to all files 
`declare(strict_types=1);`

**Review by larsroettig — COMMENTED** [review]
Looks good pls add only the static typing

File: app/code/Magento/Inventory/Indexer/IndexDataProvider.php [hunk 2]
```diff
@@ -33,13 +34,49 @@ public function __construct(ResourceConnection $resourceConnection)
...
+    }
+
+    /**
+     * Prepare select.
+     *
+     * @param int $stockId
+     * @param array $conditions
+     * @return \ArrayIterator|\Magento\Framework\DB\Select
+     */
+    private function prepareSelect($stockId, array $conditions = [])
```
Comment (ishakhsuvarov) [inline app/code/Magento/Inventory/Indexer/IndexDataProvider.php:79]: Please add type hints to the method declaration

File: app/code/Magento/Inventory/Indexer/IndexDataProvider.php [hunk 3]
```diff
@@ -76,12 +113,14 @@ public function getData(int $stockId, array $skuList = []): \ArrayIterator
             ->where('stock_source_link.' . StockSourceLink::STOCK_ID . ' = ?', $stockId)
             ->where('stock_source_link.' . StockSourceLink::SOURCE_ID . ' IN (?)', $sourceIds);
 
-        if (count($skuList) !== 0) {
-            $select->where('source_item.' . SourceItemInterface::SKU . ' IN (?)', $skuList);
+        if (!empty($conditions)) {
```
Comment (ishakhsuvarov) [inline app/code/Magento/Inventory/Indexer/IndexDataProvider.php:116]: `!empty` seems to be redundant here

File: app/code/Magento/Inventory/Indexer/IndexDataProvider.php [hunk 4]
```diff
@@ -33,13 +34,49 @@ public function __construct(ResourceConnection $resourceConnection)
...
+
+        return new \ArrayIterator($connection->fetchAll($select));
+    }
+
+    /**
+     * Prepare select.
+     *
+     * @param int $stockId
+     * @param array $conditions
+     * @return \ArrayIterator|\Magento\Framework\DB\Select
```
Comment (ishakhsuvarov) [inline app/code/Magento/Inventory/Indexer/IndexDataProvider.php:77]: It would be much more "defensive" to return just one of the types

**Pull request merged**